abci/app.go

```diff
@@ -0,0 +1,377 @@
+}
+
+// Commit the state and return the application Merkle root hash
+// Here we actually write the staged transactions into the database.
+// For details on why it has to be done here, check the Crash recovery section
+// of the ABCI spec
+func (app ForumApp) Commit(_ context.Context, commit *abci.RequestCommit) (*abci.ResponseCommit, error) {
+	if err := app.onGoingBlock.Commit(); err != nil {
+		panic(err)
+	}
+	saveState(&app.state)
+	return &abci.ResponseCommit{}, nil
+}
+
+// State Sync Connection
+// List available snapshots
+func (ForumApp) ListSnapshots(_ context.Context, listsnapshot *abci.RequestListSnapshots) (*abci.ResponseListSnapshots, error) {
+	return &abci.ResponseListSnapshots{}, nil
+}
+
+func (ForumApp) OfferSnapshot(_ context.Context, offersnapshot *abci.RequestOfferSnapshot) (*abci.ResponseOfferSnapshot, error) {
+	return &abci.ResponseOfferSnapshot{}, nil
+}
+
+func (ForumApp) LoadSnapshotChunk(_ context.Context, loadsnapshotchunk *abci.RequestLoadSnapshotChunk) (*abci.ResponseLoadSnapshotChunk, error) {
+	return &abci.ResponseLoadSnapshotChunk{}, nil
+}
+
+func (ForumApp) ApplySnapshotChunk(_ context.Context, applysnapshotchunk *abci.RequestApplySnapshotChunk) (*abci.ResponseApplySnapshotChunk, error) {
+	return &abci.ResponseApplySnapshotChunk{}, nil
+}
+
+func (app ForumApp) ExtendVote(_ context.Context, extendvote *abci.RequestExtendVote) (*abci.ResponseExtendVote, error) {
+	fmt.Println("Entered extend vote")
+
+	return &abci.ResponseExtendVote{VoteExtension: []byte(app.CurseWords)}, nil
+}
+
+func (app ForumApp) VerifyVoteExtension(_ context.Context, req *abci.RequestVerifyVoteExtension) (*abci.ResponseVerifyVoteExtension, error) {
+	fmt.Println("Entered verify extension") // Will not be called for extensions generated by this validator
+	if _, ok := app.valAddrToPubKeyMap[string(req.ValidatorAddress)]; !ok {
+		// we do not have a validator with this address mapped; this should never happen
+		panic(fmt.Errorf("unknown validator"))
+	}
+	curseWords := strings.Split(string(req.VoteExtension), "|")
+	tmpCurseWordMap := make(map[string]struct{})
+	// Verify that we do not have double words and the validator is not trying to cheat us
+	for _, word := range curseWords {
+		tmpCurseWordMap[word] = struct{}{}
+	}
+	if len(tmpCurseWordMap) < len(curseWords) {
+		// Extension repeats words
+		return &abci.ResponseVerifyVoteExtension{Status: abci.ResponseVerifyVoteExtension_REJECT}, nil
+	}
+	return &abci.ResponseVerifyVoteExtension{Status: abci.ResponseVerifyVoteExtension_ACCEPT}, nil
+}
+
+func (app *ForumApp) getWordsFromVe(voteExtensions []abci.ExtendedVoteInfo) string {
+	curseWordMap := make(map[string]int)
+	for _, vote := range voteExtensions {
+
+		// This code gets the curse words and makes sure that we do not add them more than once
+		// Thus ensuring each validator only adds one word once
+		voteExtensionString := string(vote.GetVoteExtension())
+		voteExtensionString = DedupWords(voteExtensionString)
```

Comment:
You shouldn't need to dedup here... if there are duplicates here, you should panic?

Reply:
I mean, this is called from `PrepareProposal`, and the vote extensions are the ones we gathered locally. So they should all have passed `VerifyVoteExtension`

Reply:
You are right, this would make sense in `ProcessProposal` to validate the proposer did not mess with the count, but for that to make sense we would need to add gossiping of the vote extensions..so I will remove this from here to simplofy,.